### ISSUE TITLE
Running multiple instances at once

### ISSUE THREAD

**Issue by x2b** [rsnapshot/rsnapshot#98]
I have the following problem: I am using systemd timers to automate the backups as described [here](https://wiki.archlinux.org/index.php/Rsnapshot#Automation). Since my machine is not running 24/7 I use the `Persistent` option: If some timers were not run due to the machine being switched off, then the corresponding units are immediately run after startup. 

The problem is that this leads to the {hourly,daily,monthly} instances being run at the same time. This leads to some units failing because the lockfile already exists. I am wondering whether there is a way around this problem. Maybe rsnapshot could wait for the process related to the lockfile to terminate?

**Comment by bebehei** [comment]
Yeah, I see the use-case. We had such a Pull Request, too. #57 But I rejected the code. In general, I'd merge a PR if the code quality is ok.

**Comment by sam-at-github** [comment]
Note too familiar with `systemd`, but are you sure there is not a synchronize option?  An alternative to systemd is `anacron`. It was created to run cron jobs on machines that aren't up 24/7. See man anacron, and the `-s` option.

**Comment by x2b** [comment]
Ok, this is my idea... Discuss :)

**Comment by bebehei** [comment]
Sry for replying so late. In the mail I haven't seen, that you made also a commit.

Thanks for this idea! This will be included to rsnapshot 1.5.

**Comment by temoto** [comment]
@bebehei ping on progress.

@x2b could systemd `[Unit] After` be used to control concurrency?

**Comment by x2b** [comment]
I asked the systemd guys on the IRC a couple of months ago. They told me that `After` / `Requires` does not work in this situation and that there is no way to order timers.

**Comment by nkadel-skyhook** [comment]
Then it sounds like time for anacron, or a poling wrapper to check for lock files and  the age of the last backup before running monthly, weekly, daily, or hourly, in that order.

**Comment by markstos** [comment]
@nkadel-skyhook, I can't find docs that say that `anacron` will solve the ordering problem, looking at both the `anacron` man page and also the `anacron` man page. 

@x2b There is another workaround with `systemd`: You could set `Persistent` option on one of the runs, but not all of them, to reduce the chance of conflict. So you could set daily runs to be Persistent, but not your weekly runs.   

You could also set more complex service definitions to avoid this. Perhaps this:
  
 * Create a service named `rsnapshot-daily-and-weekly.service`.  This service would run `rsnapshot daily && rsnapshot weekly`.   It's time would be set to run on the schedule of the weekly timer, while the "daily" timer would be set to not run on this day. 
 * Create a service named `rsnapshot-daily-and-weekly-and-monthly`  that runs all three backups in sequence, and update the other timers to not run on the first of the month.

I have not tested the above approach. I'm using rsnapshot+systemd on an always-on server.

**Comment by jantman** [comment]
I know this has been idle for a while, but I solved this issue on my laptop with a bit of a hack:

I have *one* systemd timer job that runs daily, and runs a shell script that wraps rsnapshot. Granted, I only do daily and weekly, no monthly, but the logic could be the same:

The shell script runs the daily job.
It then checks for a ``/root/last-weekly-rsnapshot`` file, and if it's older than 7 days, runs the weekly job and then ``touch /root/last-weekly-rsnapshot`` to update the mtime on the file.

**Comment by azrdev** [comment]
> I asked the systemd guys on the IRC a couple of months ago. They told me that After / Requires does not work in this situation and that there is no way to order timers / jobs the way we would need to here.

As I understand it, systemd does not allow (except for hacks) to have services started sequentially, with each service waiting until the previous has *finished*.
So rsnapshot should be able to [*wait*](https://linux.die.net/man/1/wait) for another instance and run as soon as that one is finished, instead of failing if there is another instance.

Independent of wether this is a systemd-issue, I think it would improve rsnapshots design: AIUI, currently you have to estimate the runtime (!) of each backup run, and calculate the differences of cronjob fire times accordingly. If you guess too small, backup runs are skipped?

**Comment by blahgeek** [comment]
How about just prefixing the command with `flock`?

**Comment by PaulGuijt** [comment]
What can go wrong if e.g. an hourly and a daily rsnapshot are made simultaneously to the same nfs mount?

**Comment by pescepalla** [comment]
I wrote a Bash script to avoid this problem. It's an ugly hack, but it works: https://pastebin.com/MSsYmdMj
